File: src/main/java/com/dm/data/repository/CourseRepository.java
package com.dm.data.repository;

import com.dm.data.entity.CourseEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

/**
 * Repository for CourseEntity.
 */
@Repository
public interface CourseRepository extends JpaRepository<CourseEntity, Long> {

    Optional<CourseEntity> findByCode(String code);

    boolean existsByCode(String code);

    List<CourseEntity> findByDepartmentId(Long departmentId);

    List<CourseEntity> findByDepartment_Faculty_Id(Long facultyId);
}

File: src/main/java/com/dm/data/repository/CourseRepository.kt
package com.dm.data.repository

import com.dm.model.Course
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository

@Repository
interface CourseRepository : JpaRepository<Course, Long> {

    fun findByCode(code: String): Course?

    fun findAllByTeacher_Id(teacherId: Long): List<Course>

    fun findAllByGroup_Id(groupId: Long): List<Course>

    fun findAllByType(type: String): List<Course>
}